compiler/optimizer/DeadTreesElimination.cpp

```diff
@@ -480,20 +485,36 @@ int32_t TR::DeadTreesElimination::performOnBlock(TR::Block *block)
    return 0;
    }
 
+typedef std::pair<TR::Node* const, TR::TreeTop* const> ReadBarToTreeTopMapEntry;
+typedef TR::typed_allocator<ReadBarToTreeTopMapEntry, TR::Region&> ReadBarToTreeTopMapAlloc;
+typedef std::map<TR::Node*, TR::TreeTop *, std::less<TR::Node*>, ReadBarToTreeTopMapAlloc> ReadBarToTreeTopMap;
+
 void TR::DeadTreesElimination::prePerformOnBlocks()
    {
    _cannotBeEliminated = false;
    _delayedRegStores = false;
 
    _targetTrees.deleteAll();
 
-   // Walk through all the blocks to remove trivial dead trees of the form
-   // treetop
-   //   => node
-   // The problem with these trees is in the scenario where the earlier use
-   // of 'node' is also dead.  However, our analysis won't find that because
-   // the reference count is > 1.
+   /*
+    * Walk through all the blocks to remove trivial dead trees in the following forms:
+    * case 1:
+    * treetop
+    *   => node
+
+    * case 2:
+    * treetop
+    *    rdbar
+    * anchor
+    *    =>rdbar
+    *
+    * The problem with these trees is in the scenario where the earlier use
+    * of 'node' is also dead.  However, our analysis won't find that because
+    * the reference count is > 1.
```

Comment:
Can/does case 2 happen for aloadi as well? What situations are resulting in this? is it sloppy ilgen or are opts specifically not caring the tree is already anchored by the anchor tree and re-anchoring it?

Reply:
I'm unclear on the last comment. It says case 2 happens when VP removes a check, however, case 2 as described in the comment talks about a check being the second reference to the rdbar. If the check was removed how did we end up in case 2 ? 

Also what exactly do you mean by "anchor" in your comment about case 2 ? 

Finally, is the following case also possible (and therefore ought to be handled):

    * anchor/check
     *    rdbar
     * treetop
     *    =>rdbar

Reply:
Sry for the confusion. I should actually remove the `check` from the comment in the code for case 2. 

- The case this code catches is the following:
```
case 2:
treetop
   rdbar
anchor
   =>rdbar
```
which is the result of VP removing the NullCHK for the following case
```
NullCHK
   rdbar
compressedrefs
   =>rdbar
```

- `anchor` means `isAnchor` query returns true which in our current code means `TR::compressedRefs`

-  this case is possible 
```
anchor/check
    rdbar
 treetop
    =>rdbar
```

It falls into existing case 1 andis handled by this existing if and doesn't need extra handling https://github.com/eclipse/omr/blob/f81a86655fae3c72e47ed45de7d66e1073913416/compiler/optimizer/DeadTreesElimination.cpp#L504-L510

Reply:
For `aloadi` this is handled in the normal dead tree elimination because for normal `aloadi` a treetop can always be removed without looking at its reference count. For `xrdbari`, we only apply normal dead tree elimination when the reference count is one (in `treeCanPossiblyBeRemoved` method). If we don't remove the first treetop in trivial dead tree elimination,  normal dead tree elimination won't kick in.

Reply:
ok so this is a result of an optimizationand not sloppy ilgen so it seems reasonable to handle this here